StardewArchipelago/Locations/CodeInjections/Modded/BoardingHouseInjections.cs

```diff
@@ -0,0 +1,138 @@
+            _archipelago = archipelago;
+            _shopReplacer = shopReplacer;
+        }
+        
+        // public override void update(GameTime time)
+        public static void Update_ReplaceDwarfShopChecks_Postfix(ShopMenu __instance, GameTime time)
+        {
+            try
+            {
+                // We only run this once for each menu
+                if (_lastShopMenuUpdated == __instance)
+                {
+                    return;
+                }
+
+                _lastShopMenuUpdated = __instance;
+                var myActiveHints = _archipelago.GetMyActiveHints();
+                ReplaceCraftsanityRecipes(__instance, myActiveHints);
+
+                __instance.forSale = __instance.itemPriceAndStock.Keys.ToList();
+                return; //  run original logic
+            }
+            catch (Exception ex)
+            {
+                _monitor.Log($"Failed in {nameof(Update_ReplaceDwarfShopChecks_Postfix)}:\n{ex}", LogLevel.Error);
+                return; // run original logic
+            }
+        }
+
+
+        private static void ReplaceCraftsanityRecipes(ShopMenu shopMenu, Hint[] myActiveHints)
+        {
+            if (!_archipelago.SlotData.Craftsanity.HasFlag(Craftsanity.All))
+            {
+                return;
+            }
+
+            foreach (var (shopIdentification, recipes) in craftsanityRecipes)
+            {
+                if (!shopIdentification.IsCorrectShop(shopMenu))
+                {
+                    continue;
+                }
+
+                foreach (var recipe in recipes)
+                {
+                    _shopReplacer.PlaceShopRecipeCheck(shopMenu.itemPriceAndStock, $"{recipe.ItemName} Recipe", recipe.ItemName, myActiveHints, recipe.Price);
+                }
+            }
+        }
+
+        public static bool CheckForAction_TreasureChestLocation_Prefix(Chest __instance, Farmer who, bool justCheckingForActivity, ref bool __result)
+        {
+            try
+            {
+                var playerLocation = Game1.player.currentLocation.Name;
+                if (justCheckingForActivity || !abandonedMines.Keys.Contains(playerLocation))
+                {
+                    return true; //run original logic
+                }
+                var currentChest = abandonedMines[playerLocation];
+                if (__instance.items.Count <= 0 || _locationChecker.IsLocationChecked(currentChest))
+                {
+                    return true; // run original logic
+                }
+                who.currentLocation.playSound("openChest");
+                if (__instance.synchronized.Value)
+                    __instance.GetMutex().RequestLock(() => __instance.openChestEvent.Fire());
+                else
+                    __instance.performOpenChest();
+                var obj = __instance.items[0];
+                __instance.items[0] = null;
+                __instance.items.RemoveAt(0);
+                __result = true;
```

Comment:
You should set the return value at the very end of the function (right before the actual return of the patch), since it's intended as the last thing that happens in the original code